File: python/201-250/207. Course Schedule.py
import collections
class Solution(object):
    def canFinish(self, n, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = {}
        for a, b in prerequisites:
            if a in graph and b not in graph[a]:
                graph[a].append(b)
            else:
                graph[a] = [b]
        remove_list = [i for i in range(n) if i not in graph] 
        while remove_list:
            cur_list = [key for key in graph]
            temp = []
            for i in remove_list:
                for key in graph:
                    if i in graph[key]:
                        graph[key].remove(i)
                        if not graph[key]:
                            temp.append(key)
                for key in temp:
                    if key in graph:
                        del graph[key]
            remove_list = [i for i in cur_list if i not in graph] 
        return len([key for key in graph]) == 0


sol = Solution()
sol.canFinish(2, [[1,0]])

File: cpp/201-250/207. Course Schedule.cc
class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<unordered_set<int>> graph = make_graph(numCourses, prerequisites);
        vector<bool> onpath(numCourses, false);
        bool res = true;
        for (int i = 0; i < numCourses; i++) {
            if (onpath[i])
                continue;
            vector<bool> visited(numCourses, false);
            bool res = dfs(i, graph, visited, onpath);
            if (!res)
                return false;
        }
        return true;
    }
private:
    vector<unordered_set<int>> make_graph(int numCourses, vector<pair<int, int>> &prerequisites) {
        vector<unordered_set<int>> graph(numCourses);
        for (auto it : prerequisites) {
            graph[it.first].insert(it.second);
        }
        return graph;
    }
    bool dfs(int node, vector<unordered_set<int>> &graph, vector<bool> &visited, vector<bool> &onpath) {
        if (onpath[node])
            return true;
        if (visited[node])
            return false;
        visited[node] = true;
        for (int i : graph[node]) {
            bool res = dfs(i, graph, visited, onpath);
            if (!res)
                return false;
        }
        onpath[node] = true;
        return true;
    }
};